### PR TITLE
feat: Support Windows 2004 with containerd + hyperv

### PR DESCRIPTION
**Pull request by AbelHu**
**Reason for Change**:

feat: Support Windows 2004 with containerd + hyperv

**Issue Fixed**:

**Credit Where Due**:

Does this change contain code from or inspired by another project?

- [ ] No
- [ ] Yes

If "Yes," did you notify that project's maintainers and provide attribution?

- [ ] No
- [ ] Yes

**Requirements**:

- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

### CODE REVIEW

File: parts/k8s/kuberneteswindowssetup.ps1 [hunk 1]
```diff
@@ -180,7 +182,7 @@ try
     # to the windows machine, and run the script manually to watch
     # the output.
     if ($true) {
-        Write-Log "Provisioning $global:DockerServiceName... with IP $MasterIP"
+        Write-Log ".\CustomDataSetupScript.ps1 -MasterIP $MasterIP -KubeDnsServiceIp $KubeDnsServiceIp -MasterFQDNPrefix $MasterFQDNPrefix -Location $Location -AgentKey $AgentKey -AADClientId $AADClientId -AADClientSecret $AADClientSecret -NetworkAPIVersion $NetworkAPIVersion -TargetEnvironment $TargetEnvironment"
```
Comment (mainred) [inline parts/k8s/kuberneteswindowssetup.ps1:197]: Should we print AADClientSecret considering the security?

File: parts/k8s/windowscontainerdfunc.ps1 [hunk 2]
```diff
@@ -21,6 +21,67 @@ function RegisterContainerDService {
...
+    [Parameter(Mandatory = $true)][string[]]
+    $builds,
+    [Parameter(Mandatory = $true)][string]
+    $image
+  )
+  
+  Write-Host "Adding hyperv runtimes $builds"
+  $hypervRuntimes = ""
+  ForEach ($buildNumber in $builds) {
+    $windowsVersion = Select-Windows-Version -buildNumber $buildNumber
```
Comment (mainred) [inline parts/k8s/windowscontainerdfunc.ps1:58]: Could we just remove `$windowsVersion = Select-Windows-Version -buildNumber $buildNumber` into createHypervisorRuntime? for less function parameters?

File: parts/windowsparams.t [hunk 3]
```diff
@@ -67,4 +67,18 @@
         "description": "The version of Docker to be installed on Windows Nodes"
       },
       "type": "string"
+    },
+    "defaultContainerdRuntimeHandler": {
+      "defaultValue": "process",
+      "metadata": {
+        "description": "The containerd handler type (process isolated or hyperv)"
```
Comment (mainred) [inline parts/windowsparams.t:74]: Could we add allowed values for `defaultContainerdRuntimeHandler`

**Review by mainred — APPROVED** [review]
Left some suggestions.

File: parts/k8s/windowscontainerdfunc.ps1 (hunk 2 above)
Comment (AbelHu) [reply parts/k8s/windowscontainerdfunc.ps1:58]: I want to keep this function to simplify the copy from master branch later.

File: parts/windowsparams.t (hunk 3 above)
Comment (AbelHu) [reply parts/windowsparams.t:74]: May I ask the reason for your question? I cannot see any potential issues.

Comment (mainred) [reply parts/windowsparams.t:74]: The reason is somebody want to know the allowed values for `defaultContainerdRuntimeHandler`, just like swagger description. Does this make sense?
My comments left don't block you from merging, just suggestions.

Comment (AbelHu) [reply parts/windowsparams.t:74]: Allowed values were added in the description. `process` or `hyperv`

Comment (AbelHu) [reply parts/windowsparams.t:74]: BTW: This template is not exposed to customers but used internally.

**Pull request merged**